aten/src/ATen/native/cuda/IndexKernel.cu

```diff
@@ -170,6 +204,22 @@ static void index_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayR
   });
 }
 
+static void index_fill_kernel(
+  TensorIterator& iter,
+  int64_t dim,
+  int64_t self_dim_size,
+  int64_t self_dim_stride,
+  Scalar source) {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+    at::ScalarType::Half, at::ScalarType::Bool, at::ScalarType::BFloat16,
+    iter.dtype(), "index_fill_cuda", [&] {
+    using dtype = OpaqueType<sizeof(scalar_t)>;
+    auto fill_val = source.to<scalar_t>();
+    auto fill_val_opaque = *reinterpret_cast<dtype*>(&fill_val);
```

Comment:
why do we need this `reinterpret_cast` here?

Reply:
To `cast` to an Opaque type. Or could I use it directly?

Reply:
I don't think you need a `reinterpret_cast` here

Reply:
I do not think you can, not all dtypes could be converted to Opaque, the conversions are simply not defined. So, the only way is through casting the pointer, not the value.